aten/src/ATen/native/mkldnn/xpu/Conv.cpp

```diff
@@ -0,0 +1,726 @@
+        params.padding,
+        params.output_padding,
+        params.dilation,
+        params.groups,
+        attr);
+  } else {
+    // oneDNN supports padding the two sides of src with different values
+    // the padding order should be front_top_left and back_bottom_right
+    auto padding_front_top_left = params.padding;
+    auto padding_back_bottom_right = params.padding;
+
+    // PyTorch constant_pad_nd:
+    // can pad different value to the two sides of Conv input (W, H, D)
+    // (padding_left, padding_right,
+    //  padding_top, padding_bottom,
+    //  padding_front, padding_back)
+    if (pad_nd.vec().size() > 0) {
+      for (int i = 0; i < dim; ++i) {
+        padding_front_top_left[i] += pad_nd[2 * dim - 2 * i - 2]; // 4, 2, 0
+        padding_back_bottom_right[i] += pad_nd[2 * dim - 2 * i - 1]; // 5, 3, 1
+      }
+    }
+
+    // create output and propagate memory format
+    if (! output_r.defined()) {
+      auto dst_tz = conv_dst_tz(
+          input.ndimension(),
+          input.sizes(),
+          weight.sizes(),
+          padding_front_top_left,
+          padding_back_bottom_right,
+          params.stride,
+          params.dilation);
+      output = at::empty(dst_tz, input.options(), mfmt);
+    }
+    output = xpu::onednn::convolution(
+        output,
+        input,
+        weight,
+        bias,
+        padding_front_top_left,
+        padding_back_bottom_right,
+        params.stride,
+        params.dilation,
+        params.groups,
+        attr);
+  }
+
+  if (ndim == 3) {
+    output = view3d(output);
+  }
+  if (output_r.defined() && !output_r.is_same(output)) {
+    output_r.copy_(output);
+  } else {
+    output_r = output;
+  }
+  return output_r;
+}
+
+Tensor _convolution(
+    const Tensor& input_r,
+    const Tensor& weight_r,
+    const Tensor& bias_r,
+    IntArrayRef stride_,
+    IntArrayRef padding_,
+    IntArrayRef dilation_,
+    bool transposed_,
+    IntArrayRef output_padding_,
+    int64_t groups_,
+    Attr attr) {
+  Tensor output_r;
+  return _convolution_out(
+      output_r,
+      input_r,
+      weight_r,
+      bias_r,
+      stride_,
+      padding_,
+      dilation_,
+      transposed_,
+      output_padding_,
+      groups_,
+      attr);
+}
+
+Tensor convolution_overrideable(
+    const Tensor& input_r,
+    const Tensor& weight_r,
+    const c10::optional<at::Tensor>& bias_r_opt,
+    IntArrayRef stride_,
+    IntArrayRef padding_,
+    IntArrayRef dilation_,
+    bool transposed_,
+    IntArrayRef output_padding_,
+    int64_t groups_) {
+  c10::MaybeOwned<Tensor> bias_r_maybe_owned =
+      at::borrow_from_optional_tensor(bias_r_opt);
+  const Tensor& bias_r = *bias_r_maybe_owned;
+  return _convolution(
+      input_r,
+      weight_r,
+      bias_r,
+      stride_,
+      padding_,
+      dilation_,
+      transposed_,
+      output_padding_,
+      groups_,
+      Attr());
+}
+
+std::tuple<Tensor, Tensor, Tensor> convolution_backward_overrideable(
```

Comment:
the overrideable APIs are for out-of-tree device.
With upstream, we should add a xpu path in convolution common dispatch path?

Reply:
Thanks for suggestion, I tried this, and we will need modify `native_functions.yaml` and codegen code. Also, some linking logic in pytorch is changed, which will need further link to `torch_xpu.so`. I think we do not want introduce such change at `native_functions.yaml` and codegen at present? Also, other out-of-tree device could override our registration by implementing their op in their repos?